### PR TITLE
main: use the interval tree for filling scope field

### LINKED ISSUE #3637: [Question] Defining new regex-based tags with scope defined by built-in `ctags` kinds
Actually, to add scope to regex-based tags, it is needed to do something as follows:

```
--regex-LANG=/(top)/\1/t,top/{scope=set}
--regex-LANG=/(sub)/\1/s,sub/{scope=ref}
```

so that scope field will be added to all found `sub` tags (`top` tags will be considered as their parent scope).

Question is, what if for `LANG` a built-in tag kind for `top` is already defined? How could I use the already available kind instead of redefining it via regex + `{scope=set}`?

For example, I would like to define some tags for the `case XXX` statement inside of C functions. Each `case` tag should belong to the function, used as scope kind. But I have not idea on how to define the scope, without having to re-define the `function` kind.

### PR DESCRIPTION
**Pull request by masatake**
~CXX~ main: insert tags to the interval table when "end:" field is set

Close #3637.

With this change, the combination of {scope=_intervaltab}{postrun}
regex flags works.

    $ cat /tmp/input.c
    int
    foo(void)
    {
            switch (x)
            {
            case A:
                    break;
            case B:
                    break;
            }
            return 0;
    }

    int
    bar(void)
    {
            switch (x)
            {
            case C:
                    break;
            case D:
                    break;
            }
            return 0;
    }

    $ cat args.ctags
    --sort=no
    --kinddef-C=H,caseHandler,case handler
    --regex-C=/^[ \t]*case[ \t]*+([a-zA-Z0-9_]+):.*/\1/H/{scope=_intervaltab}{postrun}

    $ ../../../ctags --options=NONE --options=args.ctags -o - /tmp/input.c
    foo /tmp/input.c    /^foo(void)$/;" f       typeref:typename:int
    bar /tmp/input.c    /^bar(void)$/;" f       typeref:typename:int
    A   /tmp/input.c    /^      case A:$/;"     H       function:foo
    B   /tmp/input.c    /^      case B:$/;"     H       function:foo
    C   /tmp/input.c    /^      case C:$/;"     H       function:bar
    D   /tmp/input.c    /^      case D:$/;"     H       function:bar

The scope fields for A, B, C, and D are filled well.

TODO
- [x] write about scope=interval in the man pages
- [x] rename _interval to interval
- [x] handle the case when the starting line of a tag is changed.
- [x] mline-regex-flags
- [x] optscript interface
- [x] optimize, reducing vStringNCopyS
- [x] add a case for updating lineNumber
- [x] add more description for the commit `main: delay reading all inputs when the parser just prepares its base parser`

### CODE REVIEW

File: man/ctags-optlib.7.rst.in [hunk 1]
```diff
@@ -334,6 +334,14 @@ representation. ``--list-regex-flags`` lists all the flags.
 	can be an empty string.  See the following description of
 	``scope=...`` flag about how this is useful.
 
+``postrun``
+	Match the pattern at the end of all the other parsing process;
+	built-in code written in C language, matching patters defined
+	with ``--mline-regex-<LANG>``, and matching patters defined
+	with ``--_mtable-regex-<LANG>``. Use this flag if you want to
+	access tags information extracted in the other parsing process
+	from your pattern, you may want to match the pattern at the end.
```
Comment (masatake) [inline man/ctags-optlib.7.rst.in:343]: `, you may want to match the pattern at the end.`
???

**Comment by masatake** [comment]
I added insertToIntervalTabMaybe in a commit and removed it in another commit in this pull request.
I should not do such a try-and-error in a pull request.

I must reorganize this pull request.

**Comment by masatake** [comment]
@pidgeon777  Could you review [man/ctags-optlib.7.rst.in](https://github.com/universal-ctags/ctags/pull/3678/files#diff-9c99995dada13c6e99f36547d40dea0c9c656c2f81007e9e6ce02ab670c6dfd1) ?

I added three items to our man page:

* the postrun flag,
* the scope=intervaltab flag, and
* an example of "using the interval table"

I'm not good at English. So I wonder if what I wrote makes sense.

**Comment by pidgeon777** [comment]
> @pidgeon777 Could you review [man/ctags-optlib.7.rst.in](https://github.com/universal-ctags/ctags/pull/3678/files#diff-9c99995dada13c6e99f36547d40dea0c9c656c2f81007e9e6ce02ab670c6dfd1) ?
> 
> I added three items to our man page:
> 
> * the postrun flag,
> * the scope=intervaltab flag, and
> * an example of "using the interval table"
> 
> I'm not good at English. So I wonder if what I wrote makes sense.

Later today I'll give a look 👍.

**Comment by masatake** [comment]
I found a bug about running nested subparsers. It is a stopper of merging this pull request.

**Pull request merged**